File: src/test/java/org/artembogomolova/demo/webapp/util/RestTemplateUtil.java
package org.artembogomolova.demo.webapp.util;

import java.io.File;
import java.io.InputStream;
import java.security.KeyStore;
import java.util.Objects;
import java.util.function.Supplier;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.apache.http.conn.ssl.NoopHostnameVerifier;
import org.apache.http.conn.ssl.SSLConnectionSocketFactory;
import org.apache.http.conn.ssl.TrustSelfSignedStrategy;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.ssl.SSLContextBuilder;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.boot.web.client.RestTemplateBuilder;
import org.springframework.core.io.FileSystemResource;
import org.springframework.http.client.ClientHttpRequestFactory;
import org.springframework.http.client.HttpComponentsClientHttpRequestFactory;
import org.springframework.stereotype.Service;

@Service
@Slf4j
public class RestTemplateUtil {

  private TestRestTemplate restTemplate;

  public TestRestTemplate getSecureRestTemplate(String keyStoreFile, String trustedKeystoreType, String keyStorePass, String keyPass) {
    if (restTemplate != null) {
      return restTemplate;
    }
    RestTemplateBuilder builder = new RestTemplateBuilder();
    restTemplate = new TestRestTemplate(builder, null, null, TestRestTemplate.HttpClientOption.ENABLE_REDIRECTS);
    builder.requestFactory(new ClientHttpRequestFactorySupplier(keyStoreFile, trustedKeystoreType, keyStorePass, keyPass))
        .buildRequestFactory();
    return restTemplate;
  }

  @RequiredArgsConstructor
  private static class ClientHttpRequestFactorySupplier implements Supplier<ClientHttpRequestFactory> {

    private final String keyStoreFile;
    private final String trustedKeystoreType;
    private final String keyStorePass;
    private final String keyPass;

    @Override
    public ClientHttpRequestFactory get() {
      return buildHttpRequestFactory(keyStoreFile, trustedKeystoreType, keyStorePass, keyPass);
    }

    private ClientHttpRequestFactory buildHttpRequestFactory(String keyStoreFile, String trustedKeystoreType, String keyStorePass, String keyPass) {
      SSLConnectionSocketFactory socketFactory;
      try {
        KeyStore keyStore = KeyStore.getInstance(trustedKeystoreType);

        File keyFile = new File(keyStoreFile);
        FileSystemResource fileSystemResource = new FileSystemResource(keyFile);

        InputStream inputStream = fileSystemResource.getInputStream();
        keyStore.load(inputStream, Objects.requireNonNull(keyStorePass)
            .toCharArray());

        socketFactory = new SSLConnectionSocketFactory(
            new SSLContextBuilder().loadTrustMaterial(null, new TrustSelfSignedStrategy())
                .loadKeyMaterial(keyStore, keyPass.toCharArray())
                .build(),
            NoopHostnameVerifier.INSTANCE);
      } catch (Exception e) {
        log.debug("SSL keystore exception", e);
        throw new RuntimeException(e);
      }
      CloseableHttpClient httpClient = HttpClients.custom()
          .setSSLSocketFactory(socketFactory)
          .build();

      return new HttpComponentsClientHttpRequestFactory(httpClient);
    }
  }
}

File: app/src/test/kotlin/org/artembogomolova/demo/webapp/test/context/web/RestTemplateUtil.kt
package org.artembogomolova.demo.webapp.test.context.web

import org.apache.http.conn.ssl.NoopHostnameVerifier
import org.apache.http.conn.ssl.SSLConnectionSocketFactory
import org.apache.http.conn.ssl.TrustSelfSignedStrategy
import org.apache.http.impl.client.HttpClients
import org.apache.http.ssl.SSLContextBuilder
import org.artembogomolova.demo.webapp.main.util.getLogger
import org.springframework.boot.test.web.client.TestRestTemplate
import org.springframework.boot.web.client.RestTemplateBuilder
import org.springframework.core.io.FileSystemResource
import org.springframework.http.client.ClientHttpRequestFactory
import org.springframework.http.client.HttpComponentsClientHttpRequestFactory
import java.io.File
import java.security.KeyStore
import java.util.Objects
import java.util.function.Supplier

class RestTemplateUtil {
    companion object {
        private var INSTANCE: RestTemplateUtil? = null
        fun getInstance(): RestTemplateUtil {
            if (INSTANCE == null) {
                INSTANCE = RestTemplateUtil()
            }
            return INSTANCE as RestTemplateUtil
        }
    }

    private var restTemplate: TestRestTemplate? = null

    fun getSecureRestTemplate(keyStoreFile: String?, trustedKeystoreType: String?, keyStorePass: String?, keyPass: String?): TestRestTemplate? {
        if (restTemplate != null) {
            return restTemplate
        }
        val builder = RestTemplateBuilder()
        restTemplate = TestRestTemplate(builder, null, null, TestRestTemplate.HttpClientOption.ENABLE_REDIRECTS)
        builder.requestFactory(ClientHttpRequestFactorySupplier(keyStoreFile!!, trustedKeystoreType!!, keyStorePass!!, keyPass!!))
            .buildRequestFactory()
        return restTemplate
    }


    private class ClientHttpRequestFactorySupplier(
        private val keyStoreFile: String,
        private val trustedKeystoreType: String,
        private val keyStorePass: String,
        private val keyPass: String
    ) : Supplier<ClientHttpRequestFactory> {

        override fun get(): ClientHttpRequestFactory {
            return buildHttpRequestFactory()
        }

        private fun buildHttpRequestFactory(

        ): ClientHttpRequestFactory {
            val log = getLogger(RestTemplateUtil::class.java)
            val socketFactory: SSLConnectionSocketFactory = try {
                val keyStore: KeyStore = KeyStore.getInstance(trustedKeystoreType)
                val keyFile = File(keyStoreFile)
                val fileSystemResource = FileSystemResource(keyFile)
                fileSystemResource.inputStream.use {
                    keyStore.load(
                        it, Objects.requireNonNull(keyStorePass)
                            .toCharArray()
                    )
                }
                SSLConnectionSocketFactory(
                    SSLContextBuilder().loadTrustMaterial(null, TrustSelfSignedStrategy())
                        .loadKeyMaterial(keyStore, keyPass.toCharArray())
                        .build(),
                    NoopHostnameVerifier.INSTANCE
                )
            } catch (e: Exception) {
                log.debug("SSL keystore exception", e)
                throw RuntimeException(e)
            }
            val httpClient = HttpClients.custom()
                .setSSLSocketFactory(socketFactory)
                .build()
            return HttpComponentsClientHttpRequestFactory(httpClient)
        }
    }
}